### DISCUSSION TITLE
tilt compensation

### DISCUSSION THREAD

**Discussion by vgla** [jeff-burright/Autopilot_ESP32_wifi#5] [General]
Hello Jeff,

Thank you so much for your excellent work.

I installed your code on an esp32 and it works but I'm still having issues with the tilt correction. As soon as I tilt the compass, the heading is changing by a lot (+/- 40°)! Do you have an idea how I should adapt the code? I have an LSM303/L3G 9-axis compass chip (https://www.ebay.com/itm/202091223251) and I calibrated my magnetometer using the calibration code from Polulu. I also tried to adapt the gravity value using the raw measurements from the L3G exemple (serial script).

Thank you again!

**Comment by jeff-burright** [comment]
Hi, sorry to hear it's giving you trouble. I'll take a look at it again with my prototype here at home and see if I can think why it's doing that for you. Let's rule out a couple basics first. When you calibrated the magnetometer and got a set of min/max values from the sketch, did you put those values into the pilot's main .INO file, starting at line 296? Also, are you orienting the IMU along the Z axis in its resting state (i.e., laying the chip flat on a table far away from interference sources?

The tilt compensation code is at the bottom of the Compass.ino file for reference, but you shouldn't need to touch it. The heading reading value for the pilot is set as heading=Mag_Heading_Degrees in the Subs.ino tab. This value converts the tilt compensation Mag_Heading value in the Compass.ino tab into degrees for displaying in the UI. 

If you want to experiment, you can comment out the heading=Mag_Heading_Degrees line in the Subs.ino tab and comment in heading=ToDeg(yaw). The original pilot code I built from was using the yaw variable for reasons I don't know.
